source/common/buffer/watermark_buffer.h

```diff
@@ -0,0 +1,64 @@
+#pragma once
+
+#include <string>
+
+#include "envoy/buffer/buffer.h"
+
+namespace Envoy {
+namespace Buffer {
+
+// A wrapper for an underlying buffer which does watermark validation.
+// The underlying buffer's ownership is transfered to the Watermark buffer.   Each time the inner
+// buffer is resized (written to or drained), the watermarks are checked.  As the buffer size
+// transitions from under the low watermark to above the high watermark, the above_high_watermark
+// function is called one time. It will not be called again until the buffer is drained below the
+// low watermark, at which point the below_low_watermark function is called.
```

Comment:
Matt's earlier comment re:utilities was spot on - once I started on h2 flow control I found myself copy-pasting watermark checks so pulled this out.  I *believe* it will be reusable for at least some of the codec buffers.    Either way it guards us from screwing up and missing buffer size changes changes (I missed one in ssl/connection_impl.cc before my own tests caught me) because all edits are done through the wrapper.

@htuch let me know what you think before I go too crazy with the unit tests :-)